smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java

```diff
@@ -0,0 +1,177 @@
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, null, null);
+    }
+
+    public void setMood(Mood mood, String text)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, null, text);
+    }
+
+    public void setMood(Mood mood, MoodConcretisation concretisation)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, concretisation, null);
+    }
+
+    public void setMood(Mood mood, MoodConcretisation concretisation, String text)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        MoodElement element = buildMood(mood, concretisation, text);
+        publishMood(element);
+    }
+
+    public void clearMood()
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        MoodElement element = buildMood(null, null, null);
+        publishMood(element);
+    }
+
+    private void publishMood(MoodElement moodElement)
+            throws SmackException.NotLoggedInException, InterruptedException, PubSubException.NotALeafNodeException,
+            XMPPException.XMPPErrorException, SmackException.NotConnectedException, SmackException.NoResponseException {
+
+        if (!connection().isAuthenticated()) {
+            throw new SmackException.NotLoggedInException();
+        }
+        PubSubManager pubSubManager = PubSubManager.getInstance(connection(), connection().getUser().asBareJid());
+
+        LeafNode node = pubSubManager.getOrCreateLeafNode(MOOD_NODE);
+        node.publish(new PayloadItem<>(moodElement));
+    }
+
+    private MoodElement buildMood(Mood mood, MoodConcretisation concretisation, String text) {
+        return new MoodElement(
+                new MoodElement.MoodSubjectElement(mood, concretisation),
+                text);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood) {
+        addMoodToMessage(message, mood, null);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood, MoodConcretisation concretisation) {
+        MoodElement element = buildMood(mood, concretisation, null);
+        message.addExtension(element);
+    }
+
+    public void addMoodListener(MoodListener listener) {
```

Comment:
The methods to add or remove a listener are usually thread-safe in Smack. This is not the case here.